include/proxy-wasm/context.h

```diff
@@ -106,215 +102,192 @@ class ContextBase {
-  virtual FilterStatus onUpstreamData(int data_length, bool end_of_stream);
-  enum class PeerType : uint32_t {
-    Unknown = 0,
-    Local = 1,
-    Remote = 2,
-  };
-  virtual void onDownstreamConnectionClose(PeerType);
-  virtual void onUpstreamConnectionClose(PeerType);
-  // HTTP Filter Stream Request Downcalls.
-  virtual FilterHeadersStatus onRequestHeaders(uint32_t headers);
-  virtual FilterDataStatus onRequestBody(uint32_t body_buffer_length, bool end_of_stream);
-  virtual FilterTrailersStatus onRequestTrailers(uint32_t trailers);
-  virtual FilterMetadataStatus onRequestMetadata(uint32_t elements);
-  // HTTP Filter Stream Response Downcalls.
-  virtual FilterHeadersStatus onResponseHeaders(uint32_t headers);
-  virtual FilterDataStatus onResponseBody(uint32_t body_buffer_length, bool end_of_stream);
-  virtual FilterTrailersStatus onResponseTrailers(uint32_t trailers);
-  virtual FilterMetadataStatus onResponseMetadata(uint32_t elements);
+  FilterStatus onNetworkNewConnection() override;
+  FilterStatus onDownstreamData(uint32_t data_length, bool end_of_stream) override;
+  FilterStatus onUpstreamData(uint32_t data_length, bool end_of_stream) override;
+  void onDownstreamConnectionClose(CloseType) override;
+  void onUpstreamConnectionClose(CloseType) override;
+
   // Async call response.
-  virtual void onHttpCallResponse(uint32_t token, uint32_t headers, uint32_t body_size,
-                                  uint32_t trailers);
+  void onHttpCallResponse(HttpCallToken token, uint32_t headers, uint32_t body_size,
+                          uint32_t trailers) override;
   // Grpc
-  virtual void onGrpcReceiveInitialMetadata(uint32_t token, uint32_t elements);
-  virtual void onGrpcReceiveTrailingMetadata(uint32_t token, uint32_t trailers);
-  virtual void onGrpcReceive(uint32_t token, uint32_t response_size);
-  virtual void onGrpcClose(uint32_t token, uint32_t status_code);
-
-  // Inter-VM shared queue message arrival.
-  virtual void onQueueReady(uint32_t /* token */);
-  // General stream downcall when the stream/vm has ended.
-  virtual bool onDone();
-  // General stream downcall for logging. Occurs after onDone().
-  virtual void onLog();
-  // General stream downcall when no further stream calls will occur.
-  virtual void onDelete();
-
-  virtual void error(string_view message) {
+  void onGrpcReceiveInitialMetadata(GrpcToken token, uint32_t elements) override;
+  void onGrpcReceive(GrpcToken token, uint32_t response_size) override;
+  void onGrpcReceiveTrailingMetadata(GrpcToken token, uint32_t trailers) override;
+  void onGrpcClose(GrpcToken token, GrpcStatusCode status_code) override;
+
+  void error(string_view message) override {
     std::cerr << message << "\n";
     abort();
   }
-  virtual void unimplemented() { error("unimplemented proxy-wasm API"); }
+  WasmResult unimplemented() override {
+    error("unimplemented proxy-wasm API");
+    return WasmResult::Unimplemented;
```

Comment:
This will be never reached, since `error()` terminates with `abort()`.

Reply:
Yes, but it doesn't need to.  error is a virtual function which could be replaced by some other handler or non-handler.  It could just log for example.

Reply:
Good point.